File: task-board/backend/sync.js
const HABITS_URL = process.env.HABITS_URL || '';
const HABITS_SYNC_USER = process.env.HABITS_SYNC_USER || 'aldus';

class HabitsSync {
  constructor(habitsUrl, apiKey) {
    this.habitsUrl = habitsUrl;
    this.apiKey = apiKey;
  }

  async createTodo(title, taskBoardId) {
    try {
      const resp = await fetch(`${this.habitsUrl}/api/service/todos`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'X-Api-Key': this.apiKey },
        body: JSON.stringify({ title, taskBoardId }),
      });
      if (!resp.ok) {
        console.warn(`Habits sync createTodo failed: ${resp.status}`);
        return null;
      }
      const body = await resp.json();
      return body.id || null;
    } catch (e) {
      console.warn('Habits sync createTodo error:', e.message);
      return null;
    }
  }

  async completeTodo(habitsId) {
    try {
      const resp = await fetch(`${this.habitsUrl}/api/service/todos/${habitsId}/complete`, {
        method: 'POST',
        headers: { 'X-Api-Key': this.apiKey },
      });
      if (!resp.ok) console.warn(`Habits sync completeTodo failed: ${resp.status}`);
    } catch (e) {
      console.warn('Habits sync completeTodo error:', e.message);
    }
  }

  async deleteTodo(habitsId) {
    try {
      const resp = await fetch(`${this.habitsUrl}/api/service/todos/${habitsId}`, {
        method: 'DELETE',
        headers: { 'X-Api-Key': this.apiKey },
      });
      if (!resp.ok) console.warn(`Habits sync deleteTodo failed: ${resp.status}`);
    } catch (e) {
      console.warn('Habits sync deleteTodo error:', e.message);
    }
  }
}

let instance = null;

function initHabitsSync(apiKey) {
  if (!HABITS_URL || !apiKey) {
    console.info('Habits sync disabled (HABITS_URL or API key not set)');
    return;
  }
  instance = new HabitsSync(HABITS_URL, apiKey);
  console.info(`Habits sync enabled → ${HABITS_URL} (user: ${HABITS_SYNC_USER})`);
}

function getHabitsSync() {
  return instance;
}

module.exports = { initHabitsSync, getHabitsSync, HABITS_SYNC_USER };

File: habits/backend-rust/src/sync.rs
use reqwest::Client;
use serde_json::json;
use std::env;
use tokio::fs;
use tracing::{info, warn};

#[derive(Clone)]
pub struct TaskBoardSync {
    client: Client,
    base_url: String,
    habits_user: String,
    api_key: String,
}

impl TaskBoardSync {
    pub async fn new() -> Option<Self> {
        let base_url = env::var("TASKBOARD_URL").unwrap_or_default();
        if base_url.is_empty() {
            info!("TASKBOARD_URL not set — task-board sync disabled");
            return None;
        }

        let key_path = env::var("OCTIRON_API_KEY_PATH")
            .unwrap_or_else(|_| "/root/.octiron-api-key".to_string());

        let api_key = match fs::read_to_string(&key_path).await {
            Ok(k) => k.trim().to_string(),
            Err(_) => {
                warn!("Octiron API key not found at {key_path} — task-board sync disabled");
                return None;
            }
        };

        let habits_user = env::var("TASKBOARD_HABITS_USER")
            .unwrap_or_else(|_| "aldus".to_string());

        info!("Task-board sync enabled → {base_url} (user: {habits_user})");
        Some(Self {
            client: Client::new(),
            base_url,
            habits_user,
            api_key,
        })
    }

    /// Create a task in task-board for the habits user. Returns the task-board task ID.
    pub async fn create_task(&self, title: &str, habits_id: &str) -> Option<String> {
        let url = format!("{}/api/service/tasks", self.base_url);
        let result = self
            .client
            .post(&url)
            .header("X-Api-Key", &self.api_key)
            .json(&json!({
                "ownerUsername": self.habits_user,
                "title": title,
                "habitsId": habits_id,
            }))
            .send()
            .await;

        match result {
            Ok(resp) if resp.status().is_success() => {
                let body: serde_json::Value = resp.json().await.ok()?;
                body["id"].as_str().map(|s| s.to_string())
            }
            Ok(resp) => {
                warn!("Task-board sync create_task failed: {}", resp.status());
                None
            }
            Err(e) => {
                warn!("Task-board sync create_task error: {e}");
                None
            }
        }
    }

    /// Mark a task done by its task-board ID.
    pub async fn complete_task(&self, task_board_id: &str) {
        let url = format!("{}/api/service/tasks/{}/done", self.base_url, task_board_id);
        if let Err(e) = self
            .client
            .put(&url)
            .header("X-Api-Key", &self.api_key)
            .json(&json!({}))
            .send()
            .await
        {
            warn!("Task-board sync complete_task error: {e}");
        }
    }

    /// Delete a task by its task-board ID.
    pub async fn delete_task(&self, task_board_id: &str) {
        let url = format!("{}/api/service/tasks/{}", self.base_url, task_board_id);
        if let Err(e) = self
            .client
            .delete(&url)
            .header("X-Api-Key", &self.api_key)
            .send()
            .await
        {
            warn!("Task-board sync delete_task error: {e}");
        }
    }
}
